source/extensions/filters/http/adaptive_concurrency/controller/gradient_controller.h

```diff
@@ -33,7 +35,8 @@ namespace Controller {
   GAUGE(gradient, NeverImport)                                                                     \
   GAUGE(min_rtt_calculation_active, Accumulate)                                                    \
   GAUGE(min_rtt_msecs, NeverImport)                                                                \
-  GAUGE(sample_rtt_msecs, NeverImport)
+  GAUGE(sample_rtt_msecs, NeverImport)                                                             \
+  COUNTER(race_found)
```

Comment:
This should not be a counter- an error log would be more appropriate. If you need this as a counter, it seems like something that should be maintained in a private fork and not committed upstream.

Reply:
I just removed this race_found counter, and leveraged existing counter to verify this race condition.
But i am not sure whether it should be error log though. Regardless of whether we use the double checked lock or the tryLock approach your mentioned, the [else](https://github.com/envoyproxy/envoy/blob/e90ab1cb7e6962a3cc0f8352aa6baa3132f95400/source/extensions/filters/http/adaptive_concurrency/controller/gradient_controller.cc#L122) condition will have the possibility to hit. If the behavior is expected to happen with low chance, should we still log it as error?